src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRazorTagHelperBinderPhase.cs

```diff
@@ -188,5 +255,155 @@ private bool AssemblyContainsTagHelpers(string assemblyName, IReadOnlyList<TagHe
                 return false;
             }
         }
+
+        internal class ComponentDirectiveVisitor : DirectiveVisitor
+        {
+            private IReadOnlyList<TagHelperDescriptor> _tagHelpers;
+            private string _filePath;
+            private RazorSourceDocument _source;
+
+            public ComponentDirectiveVisitor(string filePath, IReadOnlyList<TagHelperDescriptor> tagHelpers, string currentNamespace)
+            {
+                _filePath = filePath;
+
+                // We don't want to bind non-component tag helpers in a component document.
+                _tagHelpers = tagHelpers.Where(t => IsComponentTagHelperKind(t.Kind)).ToList();
+
+                for (var i = 0; i < _tagHelpers.Count; i++)
+                {
+                    var tagHelper = _tagHelpers[i];
+                    if (tagHelper.IsComponentFullyQualifiedNameMatch() ||
+                        (currentNamespace != null && IsTypeInScope(tagHelper.GetTypeName(), currentNamespace)))
+                    {
+                        // If the component descriptor matches for a fully qualified name, using directives shouldn't matter.
+                        // Also, if the type is already in scope of the document's namespace, using isn't necessary.
+                        Matches.Add(tagHelper);
+                    }
+                }
+            }
+
+            public override HashSet<TagHelperDescriptor> Matches { get; } = new HashSet<TagHelperDescriptor>();
+
+            public override string TagHelperPrefix => null;
+
+            public override void Visit(RazorSyntaxTree tree)
+            {
+                _source = tree.Source;
+                Visit(tree.Root);
+            }
+
+            public override void VisitRazorDirective(RazorDirectiveSyntax node)
+            {
+                var descendantLiterals = node.DescendantNodes();
+                foreach (var child in descendantLiterals)
+                {
+                    if (!(child is CSharpStatementLiteralSyntax literal))
+                    {
+                        continue;
+                    }
+
+                    var context = literal.GetSpanContext();
+                    if (context == null)
+                    {
+                        // We can't find a chunk generator.
+                        continue;
+                    }
+                    else if (context.ChunkGenerator is AddTagHelperChunkGenerator addTagHelper)
+                    {
+                        // Make sure this node exists in the file we're parsing and not in its imports.
+                        if (_filePath.Equals(_source.FilePath, StringComparison.Ordinal))
+                        {
+                            addTagHelper.Diagnostics.Add(
```

Comment:
Yeah. Unlike extensible directives, we've always special cased this and add/removeTagHelper/tagHelperPrefix directives in the parser as well as the PR layer. This is why these chunk generator still exist. If this needs to be cleaned up, it can be done separately.

Reply:
Ya we should do that eventually.

Reply:
can y'all log and issue please 👍

Reply:
Filed https://github.com/aspnet/AspNetCore/issues/8098